File: ApemanPlugin/src/main/kotlin/apeman_core/features_extraction/calculators/candidate/NumTypeAccessesCandidateCalculator.java
package apeman_core.features_extraction.calculators.candidate;

import apeman_core.base_entities.FeatureType;
import apeman_core.pipes.CandidateWithFeatures;
import apeman_core.utils.TypeUtils;
import com.intellij.psi.*;

import java.util.ArrayList;
import java.util.HashSet;

public class NumTypeAccessesCandidateCalculator extends AbstractNumCandidateCalculator {

    public NumTypeAccessesCandidateCalculator(ArrayList<CandidateWithFeatures> candidates) {
        super(candidates, FeatureType.NUM_TYPE_ACCESS);
    }

    @Override
    public JavaRecursiveElementVisitor createVisitor() {
        return new NumTypeAccessesCandidateCalculator.Visitor();
    }

    private class Visitor extends CandidateVisitor {

        ArrayList<HashSet<PsiType>> usedTypes;

        @Override
        protected void initCounters() {
            if (usedTypes == null) {
                usedTypes = new ArrayList<>();
            }
            usedTypes.clear();
            methodCandidates.forEach(cand -> usedTypes.add(new HashSet<>()));
        }

        @Override
        protected int getCounterForCand(int i) {
            return usedTypes.get(i).size();
        }

        @Override
        public void visitMethod(PsiMethod method) {
            super.visitMethod(method);
            if (!isInsideMethod)
                return;

            for (int i = 0; i < methodCandidates.size(); i++) {
                if (methodCandidates.get(i).getCandidate().isInCandidate()) {
                    TypeUtils.addTypesFromMethodTo(usedTypes.get(i), method);
                }
            }
        }

        @Override
        public void visitElement(PsiElement element) {
            super.visitElement(element);
            if (!isInsideMethod)
                return;

            for (int i = 0; i < methodCandidates.size(); i++) {
                if (methodCandidates.get(i).getCandidate().isInCandidate()) {
                    TypeUtils.tryAddTypeOfElementTo(usedTypes.get(i), element);
                }
            }
        }
    }
}

File: ApemanPlugin/src/main/kotlin/apeman_core/features_extraction/calculators/candidate/NumTypeAccessesCandidateCalculator.kt
package apeman_core.features_extraction.calculators.candidate

import apeman_core.base_entities.ExtractionCandidate
import apeman_core.base_entities.FeatureType
import apeman_core.pipes.CandidateWithFeatures
import apeman_core.utils.TypeUtils
import com.intellij.psi.*

import java.util.ArrayList
import java.util.HashSet

class NumTypeAccessesCandidateCalculator(candidates: List<ExtractionCandidate>
) : AbstractNumCandidateCalculator(candidates, FeatureType.NUM_TYPE_ACCESS) {

    override fun createVisitor() = Visitor()

    inner class Visitor : AbstractNumCandidateCalculator.CandidateVisitor() {

        internal var usedTypes: ArrayList<HashSet<PsiType>>? = null

        override fun initCounters() {
            if (usedTypes == null) {
                usedTypes = ArrayList()
            }
            usedTypes!!.clear()
            methodCandidates.forEach { usedTypes!!.add(HashSet()) }
        }

        override fun getCounterForCand(i: Int): Int {
            return usedTypes!![i].size
        }

        override fun visitMethod(method: PsiMethod) {
            super.visitMethod(method)
            if (!isInsideMethod)
                return

            for (i in methodCandidates.indices) {
                if (methodCandidates[i].isInCandidate) {
                    TypeUtils.addTypesFromMethodTo(usedTypes!![i], method)
                }
            }
        }

        override fun visitElement(element: PsiElement?) {
            super.visitElement(element)
            if (!isInsideMethod)
                return

            for (i in methodCandidates.indices) {
                if (methodCandidates[i].isInCandidate) {
                    TypeUtils.tryAddTypeOfElementTo(usedTypes!![i], element!!)
                }
            }
        }
    }
}
